ignite/metrics/mean_average_precision.py

```diff
@@ -0,0 +1,409 @@
+            i_category_id = 1
+            i_xmin = 2
+            i_area = 6
+            i_crowd = 7
+            g_ignore = 8
+            d_confidence = 8
+
+        self.labels = labels
+
+        super(MeanAveragePrecision, self).__init__(output_transform=output_transform, device=device)
+
+    def _check_object_area_ranges(self) -> None:
+        for area, area_range in self.object_area_ranges.items():
+            if len(area_range) != 2 or area_range[0] >= area_range[1]:
+                raise ValueError(
+                    f"object_area_ranges must be a dict associating to each key (str) a tuple of float values (a, b) "
+                    f"where a < b (got: key={area}, value={area_range})"
+                )
+
+    @reinit__is_reduced
+    def update(self, output: Tuple[torch.Tensor, torch.Tensor]) -> None:
+        y_pred_img, y_img = output
+
+        if len(y_pred_img.shape) != 2 or y_pred_img.shape[1] != 9:
+            raise ValueError(f"detections_tensor should be of size [num_detections, 9], got {y_pred_img.shape}")
+
+        if len(y_img.shape) != 2 or y_img.shape[1] != 9:
+            raise ValueError(f"ground_truths should be of size [num_detections, 9], got {y_img.shape}")
+
+        y_pred_img = y_pred_img.to(self._device)
+        y_img = y_img.to(self._device)
+
+        categories = set()
+        categorywise_y: defaultdict = defaultdict(lambda: torch.zeros(0, 9).to(self._device))
+        for y in y_img:
+            category_id = int(y[self.labels.i_category_id])
+            categories.add(category_id)
+            categorywise_y[category_id] = torch.vstack([categorywise_y[category_id], y])
+
+        categorywise_y_pred: defaultdict = defaultdict(lambda: torch.zeros(0, 9).to(self._device))
+        for y_pred in y_pred_img:
+            category_id = int(y_pred[self.labels.i_category_id])
+            categories.add(category_id)
+            categorywise_y_pred[category_id] = torch.vstack([categorywise_y_pred[category_id], y_pred])
+
+        self.category_ids.update(categories)
+
+        for category_id in categories:
+            ious = self._compute_iou(categorywise_y[category_id], categorywise_y_pred[category_id])
+            for area_rng in self.object_area_ranges:
+                self.eval_imgs[category_id, area_rng].append(
+                    self._evaluate_image(
+                        categorywise_y[category_id],
+                        categorywise_y_pred[category_id],
+                        self.object_area_ranges[area_rng],
+                        ious,
+                    )
+                )
+
+    @reinit__is_reduced
+    def reset(self) -> None:
+        self.category_ids: set = set()
+        self.eval_imgs: defaultdict = defaultdict(list)
+
+    def _compute_iou(self, y: torch.Tensor, y_pred: torch.Tensor) -> torch.Tensor:
+        y_pred = y_pred[torch.argsort(y_pred[:, self.labels.d_confidence])]
+
+        crowd = [g[self.labels.i_crowd] for g in y]
+        ious = iou(
```

Comment:
why not use the full `y` and `y_pred` instead of extracting parts of the data ? It means having `iou` in the class and not as a generic and public method.

Reply:
It improves code readability to pass only the bounding boxes. Otherwise it would become difficult to track what is being passed where, if a new person reads it.

Reply:
And for the public iou method, I can make it a static method inside MeanAveragePrecision class.